### ISSUE TITLE
Garbled goal display in Coq 8.5

### ISSUE THREAD

**Issue by ejgallego** [psteckler/ProofGeneral#70]
Using Coq 8.5 I my goals are displayed:

```
A#SPC#:#SPC#Type
```
Instead of `A : Type`.

**Comment by ejgallego** [comment]
Ok, this was due to the coq version detected incorrectly, what is the preferred way to quickly change coq versions in this new PG?

By the way, for message display, IMHO it is not necessary to rely on the version for now, I think you can match the argument of message to see what is the top-level tag and act consequently.

**Comment by ejgallego** [comment]
Right now, to switch Coq versions on the fly I have to do:

```
(setq coq-pinned-version "VER")
(setq coq-prog-name "PATH_VER")
(setq proof-prog-name coq-prog-name)
```

**Comment by psteckler** [comment]
On Wed, Jan 25, 2017 at 8:07 AM, Emilio Jesús Gallego Arias
<notifications@github.com> wrote:
> Ok, this was due to the coq version detected incorrectly, what is the preferred way to quickly change coq versions in this new PG?

I haven't worked that out yet. How does that work in vanilla PG? I can
try to emulate it.

> By the way, for message display, IMHO it is not necessary to rely on the version for now, I think you can match the argument of message to see what is the top-level tag and act consequently.

The error you saw was for escaping/unescaping spaces, which differs
between 8.5 and 8.6. I prefer to choose that strategy once, rather
than on each message. We'll have to work out to choose the strategy
again when the Coq version changes.

-- Paul

**Comment by psteckler** [comment]
On Wed, Jan 25, 2017 at 8:26 AM, Emilio Jesús Gallego Arias
<notifications@github.com> wrote:
>
> Right now, to switch Coq versions on the fly I have to do:
>
> (setq coq-pinned-version "VER")
> (setq coq-prog-name "PATH_VER")
> (setq proof-prog-name coq-prog-name)


I can write a function that takes the prog-name, and updates
everything accordingly. Would that work?

**Comment by ejgallego** [comment]
In vanilla PG just setting `coq-prog-name` and restarting the prover will make things work.

IMHO it would be easier if at every restart of the toplevel both `coq-pinned-version` and `proof-prog-name` where updated automatically, so just setting the variable and restarting would work.

**Comment by ejgallego** [comment]
As seen in #69, the printing strategy should be updated too when restarting `coqtop` somehow.

**Comment by psteckler** [comment]
More particularly, the unescaping strategy needs to change when the Coq version changes.

**Comment by psteckler** [comment]
I added an interactive command, `coq-set-prog-name`, which updates `coq-prog-name` and `proof-prog-name`.

The unescaping change works via `proof-server-init-hook`.

You need to restart after calling this function for the changes to take effect.

See if it works!

**Comment by ejgallego** [comment]
Hi @psteckler , I will test later today, thanks!

What about the version, is re-detected any time we start? Why does it have to be a command as opposed to the init code updating from a variable every time a new `coqtop` is spawn?

There would be some way to avoid having to set `coq-pinned-version` every time we switch to `trunk`?

**Comment by ejgallego** [comment]
It works well enough, thanks! If you think modifying the init logic in the lines I propose you can keep this open, otherwise closing would be fine for me.

**Comment by psteckler** [comment]
The function I wrote just setq's `coq-prog-name` and `proof-prog-name` to the new path.

Ordinarily, the second one is set to the first one in the configuration code for Emacs `coq-mode`. So just updating the first one would not give the right value to the second one, if the script file is already open. I suppose we could, instead, add a hook to automatically update `proof-prog-name` just before it starts Coq. What do you think?

I think the right way to avoid the issue with trunk would be to extend the XML protocol with a Coq or protocol version number.

**Comment by ejgallego** [comment]
The protocol version information should already appear in after the init call (in the form of a date), is it not the case?

**Comment by ejgallego** [comment]
> I suppose we could, instead, add a hook to automatically update proof-prog-name just before it starts Coq. What do you think?

Whatever it works best, I like the current command, however it seems more resilient to me that every time you are starting a new toplevel you actually reproduce all the usual steps to get PG in a consistent state (variables, versioning, etc...)

**Comment by psteckler** [comment]
I don't see a date in the response to the XML `Init` call, in either 8.5 or 8.6. 

Are you seeing that? The command `M-x proof-server-enable-logging` records the output in the `*coq-log*` buffer. If so, is the date a compilation date? If so, that's not so helpful, because that date might be long after the code release.

**Comment by psteckler** [comment]
I made the change to use a hook to set `proof-prog-name` (unless the user has configured PG to ask for that).

So now, to change the Coq version, use `M-x set-variable` on `coq-prog-name`, using a string for the path to coqtop. Then restart Coq, and things should Just Work. Oh, unless you're using trunk. :-(

**Comment by ejgallego** [comment]
Cool thanks! That is fine for me and consistent with old PG configurations AFAICT.

Maybe if trunk is detected you could just just default to the last known version?

Regarding the version, I was wrong, init doesn't return it, it seems you have to issue and INFO call.

**Comment by ejgallego** [comment]
Well, the protocol call is actually called "About".

**Comment by psteckler** [comment]
I see, that works here. 

With trunk, the version shows up as `trunk`, which doesn't help much. :-(

The protocol version as a date shows up in the next field, I'll see if I can use that.

**Comment by psteckler** [comment]
I'm not sure it's a good idea to map protocol versions to Coq versions. The Coq version could change and the protocol version stay the same. Also, it's an additional maintenance chore to maintain that mapping.

So for now, we'll have to set the pinned version when using trunk.

**Comment by ejgallego** [comment]
I am not sure I follow you here, I am not sure what the XML protocol maintainers have in mind, but certainly protocol version should take precedence over the Coq version field.

**Comment by ejgallego** [comment]
> So for now, we'll have to set the pinned version when using trunk.

I don't follow the rationale here, trunk should mean "the latest coq version", what do we gain by not going that way?

**Comment by psteckler** [comment]
> I am not sure I follow you here, I am not sure what the XML protocol maintainers have in mind, but certainly protocol version should take precedence over the Coq version field.

Yes, I should probably remove the use of the Coq version in the PG code. Right now, it's setup for using the Coq version, which is why setting the pinned version is needed.

If I can make that change, that would remove the need to set the pinned version.

Once that's done, there's no need to associate "trunk" with any Coq version, which becomes irrelevant.

**Comment by psteckler** [comment]
OK, the main use of (coq-version) in the code was to see if the version is less than Coq 8.5, in order to issue a warning. The code now makes the assumption that the version "trunk" is not less than 8.5 (not necessarily true, but a pretty safe bet).

The other use was to set the escape tokens. The code now calls About on startup, and sets the protocol version. The escape tokens are now set according to that protocol version. 

The upshot is, you should be able to set `coq-prog-name`, and, even if it's a trunk version, not have to set a pinned version.

We are now living in the best of all possible worlds!

**Comment by ejgallego** [comment]
Sounds great, thanks! This also paves the way to improve the protocol more easily. Let me test the changes and I'll close.
